### PR TITLE
Parallelize reading from remote sources

### PR DESCRIPTION
**Pull request by raghavsethi**
This changes planning such that multiple ExchangeOperators read from
HttpPageBufferClient in parallel. This improves performance in many
cases.

### CODE REVIEW

**Comment by raghavsethi** [comment]
Supersedes @pnowojski 's #6864

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 1]
```diff
@@ -547,14 +547,12 @@ private Visitor(Session session)
...
-            // Having multiple remote ExchangeClients per outputId is not acceptable for an OutputBuffer (e.g. SharedBuffer).
-            // Setting the driver instance count to 1 here:
-            // * validates that it was either unset or set to 1
-            // * pins it so that it can not later be set to a different value
-            context.setDriverInstanceCount(1);
-
             List<Type> types = getSourceOperatorTypes(node, context.getTypes());
 
+            if (!context.getDriverInstanceCount().isPresent()) {
+                context.setDriverInstanceCount(getTaskConcurrency(session));
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: I'm not familiar with that part of the code, so maybe stupid question.

Does it mean that there will be created more hash partitioning operators (`LocalExchangeSinks`) in `LocalExecutionPlanner.Visitor::visitExchange`?

Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: Sorry, I still don't understand how this code works but all I wanted to know is, whether this change increases number of hash partitioners in simple queries like
```
set session task_concurrency=16;
select avg(totalprice) from orders group by clerk;
```
for which hash partitioning in task doing aggregation was a bottleneck.

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: Yes, it does.

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java [hunk 2]
```diff
@@ -44,6 +45,7 @@
         private final PlanNodeId sourceId;
         private final ExchangeClientSupplier exchangeClientSupplier;
         private final List<Type> types;
+        private ExchangeClient exchangeClient = null;
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java]: don't initialize to java default values

**Comment by raghavsethi** [comment]
@dain I added another commit - I believe this solves a `tried to free more memory than is reserved` problem I'm seeing when I'm running the compression stuff through verifier.

If `ExchangeClient.close()` is called after the closed check in pollPage/getNextPage, but before the synchronized block in `pollPage`, we will free memory for that page twice.

**Pull request closed without merging**